File: network_secret/cisco_type9.py
"""Cisco IOS / IOS-XE type 9 password hashes.

Type 9 is a one-way hash for local users and enable secrets, written in a
config as ``username admin secret 9 $9$salt$hash``. It is scrypt with
N=16384, r=1 and p=1 over the salt's raw ASCII bytes. It replaced type 8 to
raise the memory cost of an offline attack.

Because it is one way, `decrypt` always raises. Use `check` to test a
candidate password against an existing hash.

Note the prefix: a Cisco type 9 hash starts with "$9$", the same marker
Juniper/HPE uses for its own, unrelated substitution cipher. Pick the format
by the device it came from, not by the prefix.

Reference: ios_hash_password.c in
https://github.com/CiscoDevNet/Type-6-Password-Encode

Public API:
    encrypt(plaintext: str, salt: str | None = None) -> str
    decrypt(ciphertext: str) -> str            # always raises ValueError
    check(ciphertext: str, other: str) -> tuple[str, str, bool]
"""

from __future__ import annotations

import hashlib

from . import _cisco_hash

__all__ = ["decrypt", "encrypt", "check"]

MAGIC = "$9$"
TYPE_NUMBER = 9
# scrypt cost parameters IOS uses. N=16384 with r=1 needs about 2 MB.
SCRYPT_N = 16384
SCRYPT_R = 1
SCRYPT_P = 1


def _kdf(password: bytes, salt: bytes) -> bytes:
    return hashlib.scrypt(
        password,
        salt=salt,
        n=SCRYPT_N,
        r=SCRYPT_R,
        p=SCRYPT_P,
        dklen=_cisco_hash.DIGEST_LEN,
    )


def encrypt(plaintext: str, salt: str | None = None) -> str:
    """Hash `plaintext` into a Cisco type 9 value.

    Output is non-deterministic: a fresh random salt is drawn on every call,
    so the same password produces a different hash each time. Pass `salt`
    only to reproduce a known value, as the tests do.
    """
    return _cisco_hash.hash_password(plaintext, _kdf, TYPE_NUMBER, salt)


def decrypt(ciphertext: str) -> str:
    """Always raises: type 9 is a one-way hash, not a reversible cipher."""
    raise _cisco_hash.one_way_error(TYPE_NUMBER)


def check(ciphertext: str, other: str) -> tuple[str, str, bool]:
    """Recompute `ciphertext` from the password `other` and compare.

    The salt is taken from `ciphertext`, so an equal password reproduces the
    whole string. Returns (given, recomputed, match).
    """
    return _cisco_hash.verify(ciphertext, other, _kdf, TYPE_NUMBER)

File: web/src/lib/ciphers/cisco-type9.ts
/**
 * Cisco IOS / IOS-XE type 9 password hashes.
 *
 * scrypt with N=16384, r=1, p=1 over the salt's raw ASCII bytes. It replaced
 * type 8 to raise the memory cost of an offline attack, which is also why it
 * takes a visible moment in the browser.
 *
 * Mirrors network_secret/cisco_type9.py.
 */

import { scrypt } from "scrypt-js";
import {
  DIGEST_LEN,
  hashPassword,
  oneWayError,
  verifyHash,
  type Kdf,
} from "../cisco-hash";
import type { Cipher } from "./types";

const TYPE_NUMBER = 9;
const N = 16384;
const R = 1;
const P = 1;

const kdf: Kdf = async (password, salt) =>
  Uint8Array.from(await scrypt(password, salt, N, R, P, DIGEST_LEN));

export const ciscoType9: Cipher = {
  id: "cisco-type9",
  name: "Cisco type 9",
  vendor: "Cisco IOS",
  magic: "$9$",
  tagline:
    "IOS's scrypt password hash, the successor to type 8. One-way, and deliberately memory-hard, so it takes a moment.",
  reversible: false,
  keyed: false,
  oneWay: true,
  status: "available",
  notes: [
    "scrypt with N=16384, r=1 and p=1 over a 14-character salt, giving a 32-byte digest.",
    "Memory-hard by design: it needs about 2 MB per guess, which is what makes an offline attack expensive.",
    "Careful: a Cisco $9$ and a Juniper/HPE $9$ share a prefix and nothing else. The Juniper/HPE one is a keyless substitution cipher anyone can reverse; this one is a hash.",
    "Written as: username admin secret 9 $9$salt$hash.",
  ],
  example: {
    encoded: "$9$ihSswXDbk0kaVK$o.uyR2nMrWtjMkrQwBXUR5lVuVt/KzG23rmYvshODXI",
    plaintext: "cisco123",
  },
  async encode(plaintext: string): Promise<string> {
    return hashPassword(plaintext, kdf, TYPE_NUMBER);
  },
  async decode(): Promise<string> {
    throw oneWayError(TYPE_NUMBER);
  },
  async verify(encoded: string, plaintext: string): Promise<boolean> {
    return verifyHash(encoded, plaintext, kdf, TYPE_NUMBER);
  },
};
